posthog/temporal/batch_exports/s3_batch_export.py

```diff
@@ -970,23 +982,456 @@ async def insert_into_s3_activity_from_stage(inputs: S3InsertInputs) -> RecordsC
+
+        self.current_buffer = bytearray()
+        self.pending_uploads: dict[int, asyncio.Task] = {}  # part_number -> Future
+        self.completed_parts: dict[int, CompletedPartTypeDef] = {}  # part_number -> part_info
+        self.part_counter = 1
+        self.upload_id: str | None = None
+
+        # Backpressure control
+        self.memory_semaphore = asyncio.Semaphore(max_memory_buffer // part_size)
+        self._finalized = False
+
+    async def _get_s3_client(self) -> "S3Client":
+        """Get or create the shared S3 client.
+
+        It significantly improves performance to share a single S3 client across all uploads.
+        """
+        if self._s3_client is None:
+            config: dict[str, typing.Any] = {
+                "max_pool_connections": self.max_concurrent_uploads
+                * 2,  # Increase connection pool, so to ensure we're not limited by this
+            }
+            if self.s3_inputs.use_virtual_style_addressing:
+                config["s3"] = {"addressing_style": "virtual"}
+            boto_config = AioConfig(**config)
+
+            try:
+                client_ctx = self._session.client(
+                    "s3",
+                    region_name=self.s3_inputs.region,
+                    aws_access_key_id=self.s3_inputs.aws_access_key_id,
+                    aws_secret_access_key=self.s3_inputs.aws_secret_access_key,
+                    endpoint_url=self.s3_inputs.endpoint_url,
+                    config=boto_config,
+                )
+                self._s3_client = await client_ctx.__aenter__()
+                # Store the context manager for proper cleanup
+                self._s3_client_ctx = client_ctx
+            except ValueError as err:
+                if "Invalid endpoint" in str(err):
+                    raise InvalidS3EndpointError(str(err)) from err
+                raise
+        return self._s3_client
+
+    async def finalize_file(self):
+        await self._finalize_current_file()
+        await self._start_new_file()
+
+    async def consume_chunk(self, data: bytes):
+        if self._finalized:
+            raise RuntimeError("Consumer already finalized")
+
+        self.current_buffer.extend(data)
+        self.current_file_size += len(data)
+
+        # Upload parts when buffer is full
+        while len(self.current_buffer) >= self.part_size:
+            await self._upload_next_part()
+
+    async def _upload_next_part(self):
+        """Extract a part from buffer and upload it"""
+        if not self.upload_id:
+            await self._initialize_multipart_upload()
+
+        # Extract part data
+        part_data = bytes(self.current_buffer[: self.part_size])
+        self.current_buffer = self.current_buffer[self.part_size :]
+
+        part_number = self.part_counter
+        self.part_counter += 1
+
+        # Acquire memory semaphore (blocks if too many uploads in flight)
+        await self.memory_semaphore.acquire()
```

Comment:
Isn't
```
async with self.upload_semaphore:
```
in `_upload_part_with_cleanup`, already doing this?

Reply:
Yeah, so I would keep this one then. Since it will already limit the number of new tasks created, right?

Reply:
You could even put the `release` as a callback of a task.